regparser/web/jobs/views.py

```diff
@@ -148,3 +194,172 @@ def retrieve(self, request, *args, **kwargs):
+        Side Effects
+            Runs the ``proposal_pipeline`` command.
+
+        :arg dict validated_data: Incoming data from the POST that's already
+        been validated by the serializer.
+
+        :rtype: list[str]
+        :returns: The components of the argument string in list form.
+        """
+        reg_file = RegulationFile.objects.get(
+            hexhash=validated_data["file_hexhash"])
+        # TODO: This is a total hack; we should not be storing the contents in
+        # the DB but reading the file from the filesystem. Only doing this
+        # temporarily before changing the proposal_pipeline command to work
+        # differently.
+        path = reg_file.file.storage.path(reg_file.file.name)
+        return [
+            "proposal_pipeline",
+            path,
+            eregs_site_api_url
+        ]
+
+    def post(self, request, *args, **kwargs):
+        return self.create(request, *args, **kwargs)
+
+
+class ProposalPipelineJobViewInstance(JobViewInstance):
+    queryset = ProposalPipelineJob.objects.all()
+    serializer_class = ProposalPipelineJobSerializer
+    sub_path = "proposal-pipeline/"
+
+
+class FileUploadView(mixins.ListModelMixin, mixins.CreateModelMixin,
+                     generics.GenericAPIView):
+    parser_classes = (FileUploadParser, MultiPartParser)
+    parser_classes = (MultiPartParser,)
+    serializer_class = FileUploadSerializer
+    queryset = RegulationFile.objects.all()
+    lookup_field = "hexhash"
+    size_limit = 100000000  # Arbitrary 100MB limit.
+
+    def create(self, request, *args, **kwargs):
+        """
+        Overrides the ``create`` method of ``mixins.CreateModelMixin`` in order
+        to add the file contents to the database.
+
+        Side effects
+            Alters the DB.
+
+        :arg HttpRequest request: the incoming request.
+
+        :rtype: Response
+        :returns: JSON or HTML of the information about the file (status 201),
+            or about why the file couldn't be added (status 400).
+        """
+        serialized = self.get_serializer(data=request.data)
+        serialized.is_valid(raise_exception=True)
+
+        uploaded_file = request.data["file"]
+        if uploaded_file.size > self.size_limit:
+            return Response(dict(error="File too large (%s-byte limit)." %
+                                 self.size_limit),
+                            status=status.HTTP_400_BAD_REQUEST)
+        if uploaded_file.multiple_chunks():
+            contents = b"".join(chunk for chunk in uploaded_file.chunks())
+        else:
+            contents = uploaded_file.read()
+        md = hashlib.md5(contents)
+        hexhash = md.hexdigest()
+        filename = uploaded_file.name
+        url = file_url(hexhash)
+
+        if RegulationFile.objects.filter(hexhash=hexhash).count() > 0:
+            return Response(dict(error="File already present."),
+                            status=status.HTTP_400_BAD_REQUEST)
+        else:
```

Comment:
Minor style thing, but if we're going to `if-else` and one of the branches exits early, we probably want to make the other branch also exit. This means I'd recommend something like:
```python
if something:
    return
else:
    # some other logic
    return
```
or
```python
if something:
    return
# do some other logic
return
```
over
```python
if something:
    return
else:
    # do some logic
return
```